src/main/java/org/opensearch/security/user/User.java

```diff
@@ -305,4 +342,47 @@ public boolean isServiceAccount() {
     public boolean isPluginUser() {
         return name != null && name.startsWith("plugin:");
     }
+
+    public String toSerializedBase64() {
+        String result = this.serializedBase64;
+
+        if (result == null) {
+            this.serializedBase64 = result = Base64JDKHelper.serializeObject(this);
+        }
+
+        return result;
+    }
+
+
+    void readObject(ObjectInputStream stream) throws InvalidObjectException {
+        // This object is not supposed to directly read in order to keep compatibility with older OpenSearch versions
+        throw new InvalidObjectException("Use org.opensearch.security.user.serialized.User");
+    }
+
+    @Serial
+    private static final ObjectStreamField[] serialPersistentFields = {
+            new ObjectStreamField("name", String.class),
+            new ObjectStreamField("roles", Collections.synchronizedSet(Collections.emptySet()).getClass()),
+            new ObjectStreamField("securityRoles",  Collections.synchronizedSet(Collections.emptySet()).getClass()),
+            new ObjectStreamField("requestedTenant", String.class),
+            new ObjectStreamField("attributes", Collections.synchronizedMap(Collections.emptyMap()).getClass()),
+            new ObjectStreamField("isInjected", Boolean.TYPE)
+    };
+
+    /**
+     * Creates a backwards compatible object that can be used for serialization
+     */
+    @Serial
+    private void writeObject(ObjectOutputStream out)
+            throws IOException {
+        ObjectOutputStream.PutField fields = out.putFields();
+        fields.put("name", name);
+        fields.put("roles", Collections.synchronizedSet(new HashSet<>(this.roles)));
+        fields.put("securityRoles", Collections.synchronizedSet(new HashSet<>(this.securityRoles)));
+        fields.put("requestedTenant", requestedTenant);
+        fields.put("attributes", Collections.synchronizedMap(new HashMap<>(this.attributes)));
+        fields.put("isInjected", this.isInjected);
+
+        out.writeFields();
```

Comment:
This is not particularly nice, but it is necessary to keep backwards compatibility with nodes on older OpenSearch versions. To a certain extent, I am inclined again to use a different serialization format on clusters with only new nodes in order to get rid of these `synchronizedSet()` calls.